gprofiler/metadata/py_module_version.py

```diff
@@ -0,0 +1,282 @@
+    metadata = _get_metadata(dist)
+    if metadata:
+        # The metadata should NEVER be missing the Name: key, but if it somehow
+        # does, fall back to the known canonical name.
+        return metadata.get("Name", dist.project_name)
+    return None
+
+
+# Matches the path to libpython as it may appear in /proc/[pid]/maps, e.g. " /usr/local/lib/libpython3.9.so.1.0"
+_LIBPYTHON_MAPS_PATTERN = re.compile(r"(?<=\s)/\S*libpython\S*\.so(\.\S+)?\Z")
+
+
+def _get_libpython_path(pid: int) -> Optional[str]:
+    try:
+        f = open(f"/proc/{pid}/maps")
+    except OSError:
+        return None
+
+    with f:
+        for line in f.readlines():
+            match = _LIBPYTHON_MAPS_PATTERN.search(line.strip())
+            if match is not None:
+                return match.group()
+    return None
+
+
+# Matches PY_VERSION in Python's binary
+_PY_VERSION_STRING_PATTERN = re.compile(rb"(?<=\D)(?:2\.7|3\.1?\d)\.\d\d?(?=\x00)")
+
+
+@functools.lru_cache(maxsize=128)
+def _get_python_full_version(process: Process) -> Optional[str]:
+    bin_file = _get_libpython_path(process.pid) or f"/proc/{process.pid}/exe"
+
+    # Try to extract the version string from the binary
+    try:
+        f = open(resolve_host_path(process, bin_file), "rb")
+    except OSError:
+        return None
+
+    with f:
+        for line in f.readlines():
+            match = _PY_VERSION_STRING_PATTERN.search(line)
+            if match is not None:
+                return match.group().decode()
+    return None
+
+
+# Standard library modules are identified by being under a pythonx.y dir and *not* under site/dist-packages
+_STANDARD_LIB_PATTERN = re.compile(r"/python\d\.\d\d?/(?!.*(site|dist)-packages)")
+
+
+def _get_standard_libs_version(result: Dict[str, Optional[Tuple[str, str]]], process: Process):
+    py_version = None
+
+    for path in result:
+        match = _STANDARD_LIB_PATTERN.search(path)
+        if match is None:
+            # This module is (probably) not part of the standard library
+            continue
+
+        if py_version is None:
+            py_version = _get_python_full_version(process)
+            if py_version is None:
+                # No need to continue trying if we failed
+                return None
+        # (mypy fails to understand that py_version isn't Optional at this point)
+        result[path] = ("standard-library", py_version)  # type: ignore
+
+
+@functools.lru_cache(maxsize=128)
+def _get_packages_files(process: Process, packages_path: str) -> Dict[str, Tuple[str, str]]:
+    """Return a dict of filename: (package_name, package_version) for the packages in packages_path"""
+    # Transform packages_path to be relative to /proc/[pid]/root/
+    packages_host_path = resolve_host_path(process, packages_path)
+
+    path_to_package_info = {}
+    for dist in pkg_resources.find_distributions(packages_host_path):
+        files_iter = _files_from_record(dist) or _files_from_legacy(dist)
```

Comment:
> `_files_from_legacy(dist)`

Do we know of an example package that is using legacy format?

Reply:
I'm not sure actually if I saw one or not. I definitely saw packages without the newer format info.
This code is based on pip's `show` implementation so if they choose to support it there is no reason for us not to.

Reply:
Well now I'm curious why not just call into `pip`s implementation?

Reply:
That was the approach at first, but `pip show` requires the name of the package so it's kind of the opposite of what we need, and `pip freeze` doesn't provide the information we want

Reply:
Ah ofc because our mapping is from module file to package and not a listing of packages. Ok 👍